Libraries/LibPthread/pthread.cpp

```diff
@@ -157,70 +167,233 @@ int pthread_sigmask(int how, const sigset_t* set, sigset_t* old_set)
     return 0;
 }
 
-int pthread_mutex_init(pthread_mutex_t* mutex, const pthread_mutexattr_t* attributes)
+#define PTHREAD_FUTEX_UNLOCKED 0u
+#define PTHREAD_FUTEX_LOCKED 1u
+#define PTHREAD_FUTEX_LOCKED_WITH_WAITERS 2u
+
+static int futex_lock(uint32_t& futex_addr, uint32_t futex_flags, const timespec* abstime = nullptr, bool no_wait = false)
 {
-    mutex->lock = 0;
-    mutex->owner = 0;
-    mutex->level = 0;
-    mutex->type = attributes ? attributes->type : PTHREAD_MUTEX_NORMAL;
-    return 0;
+    ASSERT(!(futex_flags & ~(FUTEX_PRIVATE_FLAG | FUTEX_CLOCK_REALTIME)));
+    for (;;) {
+        uint32_t expected = PTHREAD_FUTEX_UNLOCKED;
+        if (AK::atomic_compare_exchange_strong(&futex_addr, expected, PTHREAD_FUTEX_LOCKED, AK::MemoryOrder::memory_order_acquire)) {
+            atomic_thread_fence(AK::MemoryOrder::memory_order_release);
+#ifdef PTHREAD_DEBUG
+            dbg() << "futex_lock @ " << &futex_addr << ": acquired @ " << &futex_addr;
+#endif
+            return 0;
+        }
+        if (no_wait)
+            return EBUSY;
+        if (expected == PTHREAD_FUTEX_LOCKED_WITH_WAITERS || AK::atomic_compare_exchange_strong(&futex_addr, expected, PTHREAD_FUTEX_LOCKED_WITH_WAITERS, AK::MemoryOrder::memory_order_acquire)) {
+#ifdef PTHREAD_DEBUG
+            dbg() << "futex_lock @ " << &futex_addr << ": waiting";
+#endif
+            // NOTE: FUTEX_WAIT takes a relative timeout, so use FUTEX_WAIT_BITSET instead!
+            int saved_errno = errno;
+            int rc = futex(&futex_addr, FUTEX_WAIT_BITSET | futex_flags, PTHREAD_FUTEX_LOCKED_WITH_WAITERS, abstime, nullptr, FUTEX_BITSET_MATCH_ANY);
+            if (rc < 0) {
+                int err = errno;
+                errno = saved_errno;
+                if (err == EAGAIN) {
+#ifdef PTHREAD_DEBUG
+                    dbg() << "futex_lock @ " << &futex_addr << ": waiting aborted, try locking again";
+#endif
+                    continue;
+                }
+#ifdef PTHREAD_DEBUG
+                dbg() << "futex_lock @ " << &futex_addr << ": wait error " << err;
+#endif
+                return err;
+            }
+            atomic_thread_fence(AK::MemoryOrder::memory_order_release);
+#ifdef PTHREAD_DEBUG
+            dbg() << "futex_lock @ " << &futex_addr << ": acquired after waiting";
+#endif
+            return 0;
+        }
+    }
 }
 
-int pthread_mutex_destroy(pthread_mutex_t*)
+static int futex_unlock(uint32_t& futex_addr, uint32_t futex_flags)
 {
+    ASSERT(!(futex_flags & ~(FUTEX_PRIVATE_FLAG | FUTEX_CLOCK_REALTIME)));
+    uint32_t expected = PTHREAD_FUTEX_LOCKED;
+    if (AK::atomic_compare_exchange_strong(&futex_addr, expected, PTHREAD_FUTEX_UNLOCKED, AK::MemoryOrder::memory_order_acquire)) {
```

Comment:
I usually just use `exchange()` (w/o comparing) on the release/unlock path

Also this needs release, not acquire.